src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs

```diff
@@ -54,9 +54,9 @@ public abstract partial class PipeReader
         /// Returns a <see cref="Stream"/> that wraps the <see cref="PipeReader"/>.
         /// </summary>
         /// <returns>The <see cref="Stream"/>.</returns>
-        public virtual Stream AsStream()
+        public virtual Stream AsStream(bool leaveOpen = false)
         {
-            return _stream ?? (_stream = new PipeReaderStream(this));
+            return _stream ?? (_stream = new PipeReaderStream(this, leaveOpen));
```

Comment:
If two calls happen to AsStream with different leaveOpen values the first one was the only one that mattered.  That seems weird.

Reply:
That's a good point.  Seems like we should return a new instance if there's already a stream cached and its LeaveOpen doesn't match leaveOpen.  Alternatively there could be two fields, one for each value, but that's probably overkill.  Or maybe just cache for the leaveOpen:true case: seems rare that you'd get multiple streams for the same pipe reader/writer when disposing of the stream would make the reader/writer unusable.

Reply:
Maybe it should be the most-truthiest value?

```C#
public virtual Stream AsStream(bool leaveOpen = false)
{
    if (_stream == null)
    {
        _stream = new PipeReaderStream(this, leaveOpen);
    }
    else if (leaveOpen)
    {
        _stream.SetLeaveOpen();
    }

    return _stream;
}
```
?

Because a later leaveOpen:true suggests that someone is expecting to still use the PipeReader after disposing the Stream.

Whichever way is decided upon, there should be a test which codifies it.

Reply:
I think having a second call to AsStream(...) affect the behavior of another PipeReaderStream is a little too clever. Is saving an allocation that important here?

Reply:
I get that all the streams mutate the same underlying pipe, but mutating another stream's StreamPipeReaderOptions is a bridge too far IMHO.

Reply:
Given that it's returning a cached instance, modifying the value makes sense to me.
* The first caller either had no opinion or explicitly didn't want to own closing the PipeReader.
* The first caller didn't dispose the stream, or we'd already be in an ObjectDisposedException state.
* The second caller asserted leaveOpen: true.
* Some third caller just got broken by the second one disposing the cached/shared stream.


So maybe it doesn't make sense.

Perhaps the caching should be removed, or at least bypassed when leaveOpen is true.

Reply:
Or, since it turns out, the PipeReaderStream doesn't care at all about a disposed "state" (which sort of violates the Dispose pattern), I think it's back to "if caching, be the most truthy"

Reply:
I have changed implementation to follow "most-truthiest" strategy.